CONTRIBUTING.md

```diff
@@ -0,0 +1,232 @@
+
+  ```
+  // Check if enumeration expands directories on the current platform
+  if (GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories)
+  ```
+
+- *Add new interfaces when it makes sense for the product, not simply for unit testing*
+
+  When a class needs to be mocked (or have a subset of its behavior mocked), prefer using virtual methods instead of adding a new interface.  VFS4G uses interfaces when multiple implementations of the interface exist in the product code.
+
+- *Check for `null` using the equality (`==`) and inequality (`!=`) operators rather than `is`*
+
+  A corollary to this guideline is that equality/inequality operators that break `null` checks should not be added (see [this post](https://stackoverflow.com/questions/40676426/what-is-the-difference-between-x-is-null-and-x-null) for an example).
+
+- *Use `nameof(...)` rather than hardcoded strings*
+
+  Using `nameof` ensures that when methods/variables are renamed the logging of those method/variable names will also be updated.
+
+### C/C++
+
+- *Do not use C-style casts.  Use C++-style casts.*
+
+  C++ style casts (e.g. `static_cast<T>`) more clearly express the intent of the programmer, allow for better validation by the compiler, and are easier to search for in the codebase.
+
+- *Declare static functions at the top of source files*
+  
+  This ensures that the functions can be called from anywhere inside the file.
+
+- *Do not use namespace `using` statements in header files*
+
+  `using` statements inside header files are picked up by all source files that include the headers and can cause unexpected errors if there are name collisions.
+
+- *Prefer `using` to full namespaces in source files*
+
+  Example:
+  ```
+  // Inside MyFavSourceFile.cpp
+
+  using std::string;
+
+  // Do not use `std::` namespace with `string`
+  static string s_myString;
+  ```
+
+- *Use a meaningful prefix for "public" free functions, and use the same prefix for all functions in a given header file*
+
+   Example:
+   ```
+   // Functions declared in VirtualizationRoots.h have "VirtualizationRoot_" prefix
+   bool VirtualizationRoot_IsOnline(VirtualizationRootHandle rootHandle);
+
+   // Static helper function in VirtualizationRoots.cpp has no prefix
+   static VirtualizationRootHandle FindOrDetectRootAtVnode(vnode_t vnode, const FsidInode& vnodeFsidInode);
+   ```
+
+## Testing
+
+- *Add new unit and functional tests when making changes*
+
+  Comprehensive tests are essential for maintaining the health and quality of the product.  For more details on writing tests see [Authoring Tests](https://github.com/Microsoft/VFSForGit/blob/master/AuthoringTests.md).
+
+- *Functional tests are black-box tests and should not build against any VFS4G product code*
+
+  Keeping the code separate helps ensure that bugs in the product code do not compromise the integrity of the functional tests.
+
+### Unit Tests
+
+- *Add `ExceptionExpected` to unit tests that are expected to have exceptions*
```

Comment:
I am thinking we need to change this name because when I see it I think `ExpectedException` that NUnit has.  Maybe this should be something like `SkipBreakOnException` or `DoNotBreakOnExceptions`?  Something to think about at least.

Reply:
FWIW, I also always read this as `ExpectedException` and have to re-read it until my brain processes it correctly :)

Reply:
Filed #883 for this